src/provider/provider_os_memory.c

```diff
@@ -0,0 +1,350 @@
+        assert(0);
+        return;
+    }
+
+    *pError = TLS_last_native_error;
+    *ppMessage = Native_error_str[*pError - UMF_OS_RESULT_SUCCESS];
+}
+
+enum umf_result_t os_get_recommended_page_size(void *provider, size_t size,
+                                               size_t *pageSize) {
+    (void)size; // unused
+
+    if (provider == NULL || pageSize == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
+    umf_os_memory_provider_params_t *os_params = &os_provider->params;
+
+    if (pageSize == NULL) {
+        if (os_params->traces) {
+            fprintf(stderr, "the pageSize argument cannot be NULL\n");
+        }
+        assert(0);
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    *pageSize = sysconf(_SC_PAGE_SIZE);
+
+    return UMF_RESULT_SUCCESS;
+}
+
+enum umf_result_t os_get_min_page_size(void *provider, void *ptr,
+                                       size_t *pageSize) {
+    (void)ptr; // unused
+
+    if (provider == NULL || pageSize == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
+    umf_os_memory_provider_params_t *os_params = &os_provider->params;
+
+    if (pageSize == NULL) {
+        if (os_params->traces) {
+            fprintf(stderr, "the pageSize argument cannot be NULL\n");
+        }
+        assert(0);
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    *pageSize = sysconf(_SC_PAGE_SIZE);
+
+    return UMF_RESULT_SUCCESS;
+}
+
+enum umf_result_t os_purge_lazy(void *provider, void *ptr, size_t size) {
+    if (provider == NULL || ptr == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
+    umf_os_memory_provider_params_t *os_params = &os_provider->params;
+
+    if (madvise(ptr, size, MADV_FREE)) {
+        TLS_last_native_error = UMF_OS_RESULT_ERROR_PURGE_LAZY_FAILED;
+        if (os_params->traces) {
+            perror("lazy purging failed");
+        }
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+    return UMF_RESULT_SUCCESS;
+}
+
+enum umf_result_t os_purge_force(void *provider, void *ptr, size_t size) {
+    if (provider == NULL || ptr == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
+    umf_os_memory_provider_params_t *os_params = &os_provider->params;
+
+    if (madvise(ptr, size, MADV_DONTNEED)) {
+        TLS_last_native_error = UMF_OS_RESULT_ERROR_PURGE_FORCE_FAILED;
+        if (os_params->traces) {
+            perror("force purging failed");
+        }
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+    return UMF_RESULT_SUCCESS;
+}
+
+const char *os_get_name(void *provider) {
+    (void)provider; // unused
+    return "OS";
```

Comment:
"Linux"?

Reply:
@PatKamin - this provider would be used on multiple OS systems